src/Servers/Kestrel/test/Interop.FunctionalTests/HttpClientHttp2InteropTests.cs

```diff
@@ -514,6 +504,1095 @@ protected override bool TryComputeLength(out long length)
+        {
+            var requestReceived = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var serverResult = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var hostBuilder = new HostBuilder()
+                .ConfigureWebHost(webHostBuilder =>
+                {
+                    ConfigureKestrel(webHostBuilder, scheme);
+                    webHostBuilder.ConfigureServices(AddTestLogging)
+                    .Configure(app => app.Run(async context =>
+                    {
+                        try
+                        {
+                            var readTask = context.Request.Body.ReadAsync(new byte[11], 0, 11);
+                            requestReceived.SetResult(0);
+                            var ex = await Assert.ThrowsAsync<IOException>(() => readTask).DefaultTimeout();
+                            serverResult.SetResult(0);
+                        }
+                        catch (Exception ex)
+                        {
+                            serverResult.SetException(ex);
+                        }
+                    }));
+                });
+            using var host = await hostBuilder.StartAsync().DefaultTimeout();
+
+            var url = host.MakeUrl(scheme);
+
+            using var client = CreateClient();
+            // The client doesn't flush the headers for requests with a body unless a continue is expected.
+            client.DefaultRequestHeaders.ExpectContinue = true;
+
+            var streamingContent = new StreamingContent();
+            var request = CreateRequestMessage(HttpMethod.Post, url, streamingContent);
+            var requestTask = client.SendAsync(request);
+            await requestReceived.Task.DefaultTimeout();
+            streamingContent.Abort();
+            await serverResult.Task.DefaultTimeout();
+            await Assert.ThrowsAnyAsync<Exception>(() => requestTask).DefaultTimeout();
+
+            await host.StopAsync().DefaultTimeout();
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(SupportedSchemes))]
+        public async Task ClientReset_BeforeRequestDataEnd_ReadThrows(string scheme)
+        {
+            var requestReceived = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var serverResult = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var hostBuilder = new HostBuilder()
+                .ConfigureWebHost(webHostBuilder =>
+                {
+                    ConfigureKestrel(webHostBuilder, scheme);
+                    webHostBuilder.ConfigureServices(AddTestLogging)
+                    .Configure(app => app.Run(async context =>
+                    {
+                        try
+                        {
+                            await ReadStreamHelloWorld(context.Request.Body);
```

Comment:
I know ReadStreamHelloWorld isn't new, but I'd change it so that it reads in a loop until 0 is returned rather than assuming that "Hello World" will be received in a single read.